chips/sam4l/src/eic.rs

```diff
@@ -0,0 +1,339 @@
+// IMR: 0: The corresponding interrupt is disabled.
+//      1: The corresponding interrupt is enabled.
+// ISR: 0: An interrupt event has not occurred.
+//      1: An interrupt event has occurred.
+// ICR: Writing a one to this bit will clear the corresponding bit in ISR.
+// MODE:    0: The external interrupt is edge triggered.
+//          1: The external interrupt is level triggered.'
+// EDGE:    0: The external interrupt triggers on falling edge.
+//          1: The external interrupt triggers on rising edge.
+// LEVEL:   0: The external interrupt triggers on low level.
+//          1: The external interrupt triggers on high level.
+// FILTER:  0: The external interrupt is not filtered.
+//          1: The external interrupt is filtered.
+// ASYNC:   0: The external interrupt is synchronized to CLK_SYNC.
+//          1: The external interrupt is asynchronous.
+// EN: Writing a one to this bit will enable the corresponding external interrupt.
+// DIS: Writing a one to this bit will disable the corresponding external interrupt.
+// CTRL:    0: The corresponding external interrupt is disabled.
+//          1: The corresponding external interrupt is enabled.
+
+register_bitfields![
+    u32,
+    Interrupt [
+        INT OFFSET(0) NUMBITS(32) []
+    ],
+    // Test is not being used right now
+    Test [
+        //0: This bit disables external interrupt test mode.
+        //1: This bit enables external interrupt test mode.
+        TESTEN OFFSET(31) NUMBITS(1) [],
+
+        // Writing a zero to this bit will set the input value to INTn to zero, if test mode is enabled. 
+        // Writing a one to this bit will set the input value to INTn to one, if test mode is enabled.
+        INT OFFSET(0) NUMBITS(31) []
+    ]
+];
+
+// Page 59 of SAM4L data sheet
+const EIC_BASE: StaticRef<EicRegisters> =
+    unsafe { StaticRef::new(0x400F1000 as *const EicRegisters) };
+
+pub struct Eic<'a> {
+    registers: StaticRef<EicRegisters>,
+    enabled: Cell<bool>,
+    callbacks: [OptionalCell<&'a dyn hil::eic::Client>; 9],
+}
+
+impl<'a> hil::eic::ExternalInterruptController for Eic<'a> {
+    type Line = Line;
+
+    fn line_enable(&self, line: &Self::Line, interrupt_mode: hil::eic::InterruptMode) {
+        if !self.is_enabled() {
+            self.enable();
+        }
+
+        let regs: &EicRegisters = &*self.registers;
+
+        regs.en.write(Interrupt::INT.val(*line as u32));
+
+        self.line_configure(
+            line,
+            interrupt_mode,
+            FilterMode::FilterEnable,
+            SynchronizationMode::Synchronous,
+        );
+
+        self.line_enable_interrupt(line);
+    }
+
+    fn line_disable(&self, line: &Self::Line) {
+        if !self.is_enabled() {
+            return;
+        }
+
+        let regs: &EicRegisters = &*self.registers;
+        regs.dis.write(Interrupt::INT.val(*line as u32));
+
+        self.line_disable_interrupt(line);
+
+        self.disable();
```

Comment:
Hm.. actually, thinking about this a bit more, I think this line is a problem -- what happens if you have multiple lines enabled, and then one line chooses to disable? I think that will incorrectly turn off the entire EIC?

Ideally, this would remove the explicit `enable`/`disable` calls and use the [automated peripheral management](https://docs.tockos.org/kernel/common/peripherals/index.html) instead ([here's a SPI example](https://github.com/tock/tock/blob/master/chips/sam4l/src/spi.rs#L200)). The `after_peripheral_access` method can probably just check `regs.imr.read() == 0`

Reply:
if regs.imr.read() == 0 && regs.ctrl.read() == 0 {
       self.disable();
}

Adding this check would fix this right? Is it necessary to use automated peripheral management in this case?

Reply:
Maybe not "necessary", but architecturally it's probably a bit better. Right now this code tracks the same piece of state (is the EIC enabled) in two places: (1) in the explicit `enabled` Cell, and (2) in the EIC registers. That means that whenever code updates (2) you also have to remember to correctly update (1). With automated peripheral management, you only store the state in (2), which is architecturally cleaner and eliminates possible state synchronization bugs.

Reply:
For the most current commit, I'm simply reading the register values to determine when to enable/disable the clock. The bug is fixed. I removed the explicit enabled Cell. I tried using peripheral management method but I kind of think it's a bit not worth using it for EIC. I will look into it later.